### PR TITLE
Add extra MCV placement logic

### PR DESCRIPTION
**Pull request by NotoriousPyro**
AI MCVs now place new MCVs near new tiberium fields, away from enemies buildings and its own previous con yards. Similar to a real player.

This modifies the previous logic, to:

1. Get a new random base center that is near a resource field, outside of the centered area of a previous random con yard location. Thus making the AI build outside of previous MCV locations.
2. The normal logic then is used (which selects a random location in the area based on step 1).
3. The MaxBaseRadius around the newly selected location is checked for enemies. If none exist we continue, otherwise we return null.
4. The MaxBaseRadius around the newly selected location is checked for previous construction yards (of the same player) or MCVs. If none exist in that location we continue, otherwise we return null.

Here are a few screenshots of a tweaked AI I created to produce the additional MCVs:
image

image

image

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs [hunk 1]
```diff
@@ -204,10 +218,33 @@ void DeployMcv(IBot bot, Actor mcv, bool move)
 				return null;
 			};
 
-			var baseCenter = GetRandomBaseCenter();
+			CPos baseCenter = GetRandomBaseCenter(distanceToBaseIsImportant);
```
Comment (teinarss) [inline OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs]: use `var` here.

File: OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs [hunk 2]
```diff
@@ -204,10 +218,33 @@ void DeployMcv(IBot bot, Actor mcv, bool move)
...
+			CPos? bc = findPos(baseCenter, baseCenter, Info.MinBaseRadius,
 				distanceToBaseIsImportant ? Info.MaxBaseRadius : world.Map.Grid.MaximumTileSearchRange);
+
+			if (!bc.HasValue)
+				return null;
+
+			baseCenter = bc.Value;
+
+			WPos wPos = world.Map.CenterOfCell(bc.Value);
+			WDist newBaseRadius = new WDist(Info.MaxBaseRadius * 1024);
```
Comment (teinarss) [inline OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs]: Same here.

File: OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs [hunk 3]
```diff
@@ -51,13 +51,27 @@ public class McvManagerBotModuleInfo : ConditionalTraitInfo
...
-		public CPos GetRandomBaseCenter()
+		public CPos GetRandomBaseCenter(bool distanceToBaseIsImportant)
 		{
+			if (distanceToBaseIsImportant == false)
+				return initialBaseCenter;
+
+			var tileset = world.Map.Rules.TileSet;
+			BitArray resourceTypeIndices;
+
+			resourceTypeIndices = new BitArray(tileset.TerrainInfo.Length);
```
Comment (teinarss) [inline OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs]: You can join these two `var resourceTypeIndices = new BitArray(tileset.TerrainInfo.Length);`

**Review by teinarss — CHANGES_REQUESTED** [review]

**Comment by NotoriousPyro** [comment]
Ah, yes I seem to have missed that as part of the coding standards. I will make these changes later.

File: OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs [hunk 4]
```diff
@@ -51,13 +51,26 @@ public class McvManagerBotModuleInfo : ConditionalTraitInfo
 
 	public class McvManagerBotModule : ConditionalTrait<McvManagerBotModuleInfo>, IBotTick, IBotPositionsUpdated, IGameSaveTraitData
 	{
-		public CPos GetRandomBaseCenter()
+		public CPos GetRandomBaseCenter(bool distanceToBaseIsImportant)
 		{
+			if (distanceToBaseIsImportant == false)
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs]: This can simplify to `if (!distanceToBaseIsImportant)`.

**Comment by NotoriousPyro** [comment]
I have added that change, and also a change to make the MCV placement avoid any enemy units not only buildings, as the AI was trying to place MCVs next to squads.

**Comment by NotoriousPyro** [comment]
I think I may have accidentally included a merge from one of my other branches into this PR...

**Comment by NotoriousPyro** [comment]
I'll reopen this commit when I've fixed it up in a separate branch.

**Pull request closed without merging**